### ISSUE TITLE
HMAC formats supporting the other ordering of key and data

### ISSUE THREAD

**Issue by solardiz** [openwall/john#4259]
HMAC is defined as operating on key and data, not on password and salt, and its direct uses for password hashing are actually misuses. The way we support cracking of HMAC-* in JtR is treating password as key and salt as data. This appeared to match the demand, but now we got a request to support the other ordering of key and data as well:

https://www.openwall.com/lists/john-users/2020/05/05/3

Incidentally, if #1903 is implemented it'd provide such support as a side-effect.

**Comment by magnumripper** [comment]
I thought I had added that years ago... this should be about as trivial as it ever gets. I'll have a look.

BTW I often see slightly problematic results of the HMAC formats being so promiscuous due to not having a format tag. Perhaps we could add a tag for them? Perhaps even mimic what (I think) the dynamic formats do in some situations: Be "promiscuous" if `--format=hmac-whatever` is used, but otherwise require the tag.

**Comment by solardiz** [comment]
Yes, I think we should start requiring a tag unless `--format=hmac-whatever` is used. While at it, we might also want to support arbitrary binary (hex?) salts for these. Right now, we can't have a colon or a linefeed in the salt.

**Comment by magnumripper** [comment]
Apparently that user needs iterated HMAC (perhaps actually PBKDF2) so maybe we don't need these formats after all. I think hashcat supports them though, and it should be trivial (if it wasn't for optimizations and SIMD it would basically be a one minute job). We'll see where things end up.
